### ISSUE TITLE
New rule request: Local variable might be referenced before assignment

### ISSUE THREAD

**Issue by Garrett-R** [astral-sh/ruff#6902]
It would be awesome if Ruff could catch an error (like PyCharm does) for code like this:

```python
def func(x: bool) -> int:
    if x:
        y = 1
    return y
```

**Comment by charliermarsh** [comment]
Makes sense. It's not trivial to fix because it requires supporting full branch analysis, but it would be good to catch of course.

**Comment by Garrett-R** [comment]
Yeah, seems really complicated, as does this linter magic stuff haha

**Comment by inoa-jboliveira** [comment]
At least pycharm and possibly other IDEs static code analysis can catch this bug. I was also hoping to also find this rule in Ruff.
I thought F823 would catch this, but it doesn't.

**Comment by clemente0731** [comment]
SAME ISSUE FOR "Local variable might be referenced before assignment"

**Comment by clemente0731** [comment]
https://github.com/google/pytype  can work well for this @charliermarsh

**Comment by Garrett-R** [comment]
Closing the issue — this seems to be covered by [Ty](https://github.com/astral-sh/ty)'s `possibly-unresolved-reference`!  🎊 

With this config:
```py
[tool.ty.rules]
possibly-unresolved-reference = "error"
```

and running on the above example, you get:
```py
error[possibly-unresolved-reference]: Name `y` used when possibly not defined
 --> del.py:4:12
  |
2 |     if x:
3 |         y = 1
4 |     return y
  |            ^
  |

```

**Comment by inoa-jboliveira** [comment]
This seems like way more suitable for a linter than a type checker (I know I know, there's a lot of overlap).

IIUC, since both projects share most of the code base, why not add this to a ruff rule, especially since ty isn't production ready.

**Comment by Garrett-R** [comment]
I guess I could see it in both.  Linter could certainly catch that, or a typing system since the symbol at that point has an ill-defined type, so it's at least the type checker's problem too.

> why not add this to a ruff rule, especially since Ty isn't production ready

Probably best to decide where it belongs long-term and roll with that, rather than letting the temporary Alpha state of Ty be a deciding factor on where it should live.
